File: android/src/com/heroes3/livewallpaper/SettingsActivity.java
package com.heroes3.livewallpaper;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import static android.content.Intent.CATEGORY_OPENABLE;
import static android.content.Intent.EXTRA_LOCAL_ONLY;

public class SettingsActivity extends AppCompatActivity {

    private static final int PICK_FILE_RESULT_CODE = 1000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.settings_activity);
        getSupportFragmentManager()
            .beginTransaction()
            .replace(R.id.settings, new SettingsFragment())
            .commit();
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        SharedPreferences prefs = getSharedPreferences("wallpaper_preferences", Context.MODE_PRIVATE);

        String updateFrequency = prefs.getString("update_frequency", null);
        System.out.println("FREQ: " + updateFrequency);
    }

    protected void handleLodFileSelect(Intent data) {
        Uri uri = data.getData();
        if (uri == null) {
            return;
        }

        String uriPath = uri.getPath();
        if (uriPath == null) {
            return;
        }

        boolean isCorrectFile = uriPath.toLowerCase().endsWith("h3sprite.lod");
        if (!isCorrectFile) {
            Toast
                .makeText(
                    this,
                    "Incorrect file selected",
                    Toast.LENGTH_LONG
                )
                .show();
            return;
        }

        String filePath = uriPath.substring(uriPath.indexOf(":") + 1);
        System.out.println("SELECTED FILE: " + filePath);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        System.out.println("ON ACTIVITY RESULT: " + requestCode + " : " + resultCode);
        if (/*requestCode == PICK_FILE_RESULT_CODE && */resultCode == RESULT_OK) {
            handleLodFileSelect(data);
        }
    }

    public static class SettingsFragment extends PreferenceFragmentCompat {
        @Override
        public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey);

            Preference selectLodItem = findPreference("select_lod");
            if (selectLodItem != null) {
                selectLodItem.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                    public boolean onPreferenceClick(Preference preference) {
                        Intent intent = new Intent(Intent.ACTION_GET_CONTENT)
                            .setType("application/octet-stream")
                            .addCategory(CATEGORY_OPENABLE)
                            .putExtra(EXTRA_LOCAL_ONLY, true);

                        startActivityForResult(
                            Intent.createChooser(intent, "Select file"),
                            PICK_FILE_RESULT_CODE
                        );
                        return true;
                    }
                });
            }
        }
    }
}

File: android/src/com/heroes3/livewallpaper/android/SettingsActivity.kt
package com.heroes3.livewallpaper.android

import android.app.Activity
import android.content.*
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.badlogic.gdx.utils.GdxNativesLoader
import com.heroes3.livewallpaper.R
import com.heroes3.livewallpaper.parser.AssetsParser
import java.io.*
import kotlin.concurrent.thread

class SettingsActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.settings_activity)
        supportFragmentManager
            .beginTransaction()
            .replace(R.id.settings, SettingsFragment())
            .commit()
        actionBar?.setDisplayHomeAsUpEnabled(true)
    }

    class SettingsFragment : PreferenceFragmentCompat() {
        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.root_preferences, rootKey)
        }

        override fun onPreferenceTreeClick(preference: Preference?): Boolean {
            if (preference?.key == "select_file") {
                val intent = Intent(Intent.ACTION_GET_CONTENT)
                    .setType("application/octet-stream")
                    .addCategory(Intent.CATEGORY_OPENABLE)
                    .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                    .setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY)
                    .putExtra(Intent.EXTRA_LOCAL_ONLY, true)

                startActivityForResult(
                    Intent.createChooser(intent, getString(R.string.assets_select_file_activity_title)),
                    PICK_FILE_RESULT_CODE
                )

                return true
            }

            return super.onPreferenceTreeClick(preference)
        }

        override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
            super.onActivityResult(requestCode, resultCode, intent)

            if (requestCode == PICK_FILE_RESULT_CODE
                && resultCode == Activity.RESULT_OK
                && intent?.data != null) {
                handleFileSelection(intent.data!!)
            }
        }

        private fun applyPreference(name: String, block: (parsingStatus: Preference) -> Unit) {
            val preference = findPreference<Preference>(name)
            requireActivity().runOnUiThread { preference?.apply(block) }
        }

        private fun handleFileSelection(filePath: Uri) {
            GdxNativesLoader.load()

            thread {
                var stream: InputStream? = null

                try {
                    applyPreference("select_file") {
                        it.summary = "Parsing..."
                        it.isSelectable = false
                    }

                    stream = requireContext().contentResolver.openInputStream(filePath)!!
                    AssetsParser(stream)
                        .parseLodToAtlas(
                            requireContext().filesDir.resolve("assets/sprites/test/"),
                            "assets"
                        )
                    applyPreference("select_file") { it.summary = "Parsing successfully done!" }
                } catch (e: Exception) {
                    applyPreference("select_file") { it.summary = "Something went wrong!\n${e.message}" }
                } finally {
                    stream?.close()
                    applyPreference("select_file") { it.isSelectable = true }
                }
            }
        }

        companion object {
            const val PICK_FILE_RESULT_CODE = 1
        }
    }
}